core/servicedata/service.go

```diff
@@ -73,6 +81,80 @@ func (s Service) CreateKey(ctx context.Context, key Key) (Key, error) {
 	// TODO find way to use transaction for spicedb
 	ctx = s.repository.WithTransaction(ctx)
 
+	createdServiceDataKey, err := s.createKey(ctx, key, currentUser)
+	if err != nil {
+		return Key{}, err
+	}
+
+	if err := s.repository.Commit(ctx); err != nil {
+		return Key{}, err
+	}
+
+	return createdServiceDataKey, nil
+}
+
+func (s Service) Upsert(ctx context.Context, servicedata ServiceData) (ServiceData, error) {
+	if servicedata.Key.Key == "" {
+		return ServiceData{}, ErrInvalidDetail
+	}
+
+	currentUser, err := s.userService.FetchCurrentUser(ctx)
+	if err != nil {
+		return ServiceData{}, err
+	}
+
+	project, err := s.projectService.Get(ctx, servicedata.Key.ProjectID)
+	if err != nil {
+		return ServiceData{}, err
+	}
+	servicedata.Key.ProjectID = project.ID
+	servicedata.Key.ProjectSlug = project.Slug
+
+	servicedata.Key.URN = servicedata.Key.CreateURN()
+
+	ctx = s.repository.WithTransaction(ctx)
+
+	res, err := s.resourceService.GetByURN(ctx, servicedata.Key.URN)
+	if err != nil {
+		switch {
+		case errors.Is(err, resource.ErrNotExist):
+			// create service data key if resource not exist
+			_, err := s.createKey(ctx, servicedata.Key, currentUser)
+			if err != nil {
+				return ServiceData{}, err
+			}
```

Comment:
if key is not exist then we should return not found.
the upsert only works for service data value only